requestsignature/requestsignature.go

```diff
@@ -0,0 +1,328 @@
+package requestsignature
+
+import (
+	"crypto/hmac"
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net/http"
+	"path"
+	"reflect"
+	"regexp"
+	"sort"
+	"strings"
+	"time"
+)
+
+type RequestSigner interface {
+	ValidateSignedRequest(req *http.Request) error
+}
+
+type RequestSignatureDto struct {
+					EventType string `json:"type"`
+					TenantId  string `json:"tenantId"`
+					BaseUri   string `json:"baseUri"`
+				}
+
+// the DvelopLifeCycleEventPath is path of an app endpoint, that apps must be provide
+const DvelopLifeCycleEventPath = "dvelop-cloud-lifecycle-event"
+
+// header who contains relevant headers for signature
+const signatureHeaderKey = "x-dv-signature-headers"
+
+// allowed content-type header value
+const validContentTypeHeaderValue = "application/json"
+
+// valid time differenz of request
+const timeDiff = 5 * time.Minute
+
+// Validate signature of request i.e. for validate HTTP events from cloud center
+// The middleware "HandleSignMiddleware" checks the signature of incoming requests. This is important for
+// cloud center to app authentication. The cloudcenter make an POST request to app with a signature. The middleware
+// checks if request is a POST request and the content-type header is set to "application/json".
+// If the requested signature is valid, then your handler is invoke to handle the request. If the
+// signature is invalid, the middleware returns the HTTP error 403 "Forbidden" and log the reason to your application log.
+//
+// The parameter for the "appSecret" is the base64 decoded app secret string of your app as byte array.
+//
+// More information about signature algorithm please visit the following documentation:
+// 	https://developer.d-velop.de/documentation/ccapi/en/cloud-center-api-199623589.html
+//
+// Example:
+//	func main() {
+//		// replace `Rg9iJXX0Jkun9u4Rp6no8HTNEdHlfX9aZYbFJ9b6YdQ=` with your app secret (base64-string)
+//		myAppSecret, err := base64.StdEncoding.DecodeString(`Rg9iJXX0Jkun9u4Rp6no8HTNEdHlfX9aZYbFJ9b6YdQ=`)
+//		if err != nil {
+//			panic(err)
+//		}
+//		mux := http.NewServeMux()
+//		// the path must a ressource for dvelop-cloud-lifecycle-event
+//		path := "/app/dvelop-cloud-lifecycle-event"
+//		mux.Handle(path, requestsignatur.HandleSignaturValidation(myAppSecret, time.Now)(eventHandler()))
+//	}
+//
+//	func eventHandler() http.Handler {
+//		return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+//			eventDto := new(requestsignature.RequestSignatureDto)
+//			err := json.NewDecoder(req.Body).Decode(eventDto)
+//			if err != nil {
+//				log.Print(err)
+//				http.Error(w, http.StatusText(http.StatusInternalServerError), http.StatusInternalServerError)
+//				return
+//			}
+//			doSomeStuff(eventDto)
+//		})
+//	}
+
+func HandleSignMiddleware(appSecret []byte, timeNow func() time.Time) func(http.Handler) http.Handler {
```

Comment:
Does the timeNow function have to be public or can it be an implemenation detail. For testability you have to change your package scope to be an internal test of the package.